### DISCUSSION TITLE
Idiomatic answer filtering

### DISCUSSION THREAD

**Discussion by yrashk** [LogtalkDotOrg/logtalk3#173] [Q&A]
Is there any idiomatic way (or a tried technique) for specifying filters for answers for certain predicates in particular objects?

Suppose I have some predicate that uses other non-deterministic predicates defined across my codebase that return zero or more answers. This generates all possible configuration parameters for building packages (different compilers, build tools, library artifacts, etc.). Applying all these answers is helpful in some cases as it generates the entire matrix. However, more often than not, we need to limit it to certain answers to narrow it down.

Let's say we have a predicate `compiler::c_compiler(Compiler)`, and we want to ensure that the code that uses this predicate only gets answers that are `clang`, something [superficially] similar to this:

```
with_filter(compiler::c_compiler(clang), our_goal)
```

I can probably do a lot of it manually, rewriting each such predicate to have a "filtering" version of the predicate instead of the real one (that would compare its notes with the asserted filter [hand-wavy here]) and having the actual predicates defined in its "internal" version. Maybe even with some help of term expansion. But before I do so, I just wonder if such a case has been worked on before and maybe there's something to learn from.

As always, thank you very much!

P.S. I hope my sleep deprivation is not affecting the quality of this question too much :)

**Comment by pmoura** [comment]
As you mentioned, one of the solutions is to have additional, auxiliary predicates, with some filter argument(s). As the filter always needs to be specified somewhere, a possible alternative is to use parametric objects where the parameter(s) are either variables (at message sending time) or found to the filter(s). Something like:

```logtalk
:- object(config(_Compiler_), ...).

    satisfy(...) :-
        compiler::c_compiler(_Compiler_),
        ...

:- end_object.
```

This could then be used as either `config(_)::satisfy(...)` or `config(clang)::satisfy(...)`.

Both solutions have the same expressive power but it's not clear without more details which one would be preferable from a software engineering perspective. The parametric object solution may work best when the filter(s) apply to multiple predicates.

**Reply by yrashk** [reply]
Thank you. I think my main issue here is that I want to be able to address a few properties here: 

* (a must) being able to specify multiple filters simultaneously (I want this compiler, this make, etc.)
* (a great to have) being able to specify filters as a goal to filter by an arbitrary condition

I suspect I will have to use dynamic facts to establish filters and clear them when done.

Thinking about the approach I outlined in my original message, I wish there was a way to use `alias` on predicates from the same object so that it would be easier to write a term expander to "override" the predicate with the "filtering" version. Anything you can recommend for this? The only workaround I can think of here is to introduce an approach where the object with the predicate is implemented as a concept, and then the overriding happens in the object, where we can simply call the inherited version and filter it:

```logtalk
:- category(c).

:- public(prop/1).

prop(1).
prop(2).

:- end_category.

:- object(o, imports(c)).

prop(X) :- ^^prop(X),
  X \= 2 % some filtering here.

:- end_object.
```

It seems to be a clean approach but it does require a ceremony (having a separate category)

**Comment by pmoura** [comment]
With parametric objects/categories based solution, you would likely use a parameter per filter. You could then provide common configurations (with fixed compiler, make, ... tools) by extending the parametric object and bounding the parameters. Something like:

```logtalk
:- object(config(_Compiler_, _Make_, ...), ...).
        ...
:- end_object.

:- object(gnu_config, extends(config(gcc, gmake, ...))).
        ...
:- end_object.
```

Regarding aliases, have you already read:

https://logtalk.org/2020/01/08/object-and-predicate-aliases.html

**Reply by yrashk** [reply]
I've read it indeed, but aliases do not seem to work for the "current" object:

```logtalk
:- object(o).

:- public(a/1).

a(1).

:- alias(o, [a/1 as b/1]).

:- end_object.
```

Gets `reference_error(entity_identifier,o)`, which I understand, as the object is not defined yet!

As for the parametric solution, I think this is limiting as it's not an all-or-none in my case. It is more of a "apply whichever filters you want and leave the rest", and the body of filterable properties will keep growing over time.

**Reply by pmoura** [reply]
Indeed the first argument of an `alias/2` directive must be an entity referenced in the opening directive other than the entity being defined.

**Reply by yrashk** [reply]
I wish it was possible to alias from the same object so that there would be no need to introduce the category ceremony. But I understand that this may be difficult or not even feasible or desired.

**Reply by pmoura** [reply]
That's not what the `alias/2` directive is about. But if the predicates are contained in an (imported) category, you can use `alias/2` directives in any object importing the category.

**Reply by yrashk** [reply]
Yeah, this is once again coming to the only solution I found so far: defining filterable predicates in categories. Which does introduce the category introduction ceremony. I'll try and see if I can live with it.
